src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java

```diff
@@ -0,0 +1,303 @@
+package com.intuit.graphql.orchestrator.visitors.queryVisitors;
+
+import com.intuit.graphql.orchestrator.utils.GraphQLUtil;
+import graphql.ExecutionInput;
+import graphql.GraphQLException;
+import graphql.language.Argument;
+import graphql.language.AstPrinter;
+import graphql.language.AstTransformer;
+import graphql.language.BooleanValue;
+import graphql.language.Definition;
+import graphql.language.Directive;
+import graphql.language.DirectivesContainer;
+import graphql.language.Document;
+import graphql.language.Field;
+import graphql.language.FragmentDefinition;
+import graphql.language.FragmentSpread;
+import graphql.language.InlineFragment;
+import graphql.language.Node;
+import graphql.language.OperationDefinition;
+import graphql.language.Selection;
+import graphql.language.SelectionSet;
+import graphql.language.SelectionSetContainer;
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+import lombok.Builder;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.intuit.graphql.orchestrator.utils.DirectivesUtil.DEFER_DIRECTIVE_NAME;
+import static com.intuit.graphql.orchestrator.utils.DirectivesUtil.DEFER_IF_ARG;
+import static com.intuit.graphql.orchestrator.utils.IntrospectionUtil.__typenameField;
+import static graphql.util.TreeTransformerUtil.changeNode;
+
+/*
+* Purpose of class: traverse query and create forked ei whenever there is a valid defer directive
+* */
+@Builder
+public class DeferDirectiveQueryVisitor extends QueryVisitorStub {
+
+    //constants
+    public static final String GENERATED_EIS = "GENERATED_EIS";
+    private static final AstTransformer AST_TRANSFORMER = new AstTransformer();
+
+    //variables that will be manipulated throughout
+    private final List<ExecutionInput> generatedEIs =  new ArrayList<>();
+    private final Set<String> neededFragmentSpreads = new HashSet<>();
+
+    //variables that are set when visitor is built
+    private ExecutionInput originalEI;
+    private final Document rootNode;
+    private final Map<String, FragmentDefinition> fragmentDefinitionMap;
+    private final PruneChildDeferSelectionsModifier PRUNE_CHILD_MODIFIER = PruneChildDeferSelectionsModifier
+            .builder()
+            .nestedDefersAllowed(true)
+            .build();
+
+
+    /*
+    * Functions:
+    *    Updates the field if it contains defer directive.
+    *    If it is a valid deferred node, generate new EI
+    */
+    @Override
+    public TraversalControl visitField(Field node, TraverserContext<Node> context) {
+        if(containsValidDeferDirective(node)) {
+            Node prunedNode = pruneDeferDirective(node);
+
+            if(isValidDeferredSelection(node)) {
```

Comment:
just curious if this validation should be run first before `pruneDeferDirective(node)`?

Reply:
regardless of if the validation, the directive needs to be removed. The reason being, if there is a nested field later on the defer directive should be removed or else the query to the downstream will have the directive.